### ISSUE TITLE
[feature/netcore] NullReferenceException

### ISSUE THREAD

**Issue by genusP** [OData/WebApi#1134]
I tested a build from https://github.com/robward-ms/WebApi/tree/NetCore-Latest.
My project use asp.net core 2.0.
If I add OData attribute routing, raise exception
```
Satellite.Host> System.NullReferenceException: Object reference not set to an instance of an object.
Satellite.Host>    at Microsoft.AspNet.OData.Adapters.WebApiRequestMessage.get_RouteData()
Satellite.Host>    at Microsoft.AspNet.OData.Routing.Conventions.AttributeRoutingConvention.SelectActionImpl(IWebApiControllerContext controllerContext)
Satellite.Host>    at Microsoft.AspNet.OData.Routing.Conventions.AttributeRoutingConvention.SelectAction(RouteContext routeContext)
Satellite.Host>    at Microsoft.AspNet.OData.Routing.ODataActionSelector.SelectCandidates(RouteContext context)
Satellite.Host>    at Microsoft.AspNetCore.Mvc.Internal.MvcRouteHandler.RouteAsync(RouteContext context)
Satellite.Host>    at Microsoft.AspNetCore.Routing.Route.OnRouteMatched(RouteContext context)
Satellite.Host>    at Microsoft.AspNetCore.Routing.RouteBase.RouteAsync(RouteContext context)
Satellite.Host>    at Microsoft.AspNetCore.Routing.RouteCollection.<RouteAsync>d__9.MoveNext()
Satellite.Host> --- End of stack trace from previous location where exception was thrown ---
Satellite.Host>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
Satellite.Host>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Satellite.Host>    at Microsoft.AspNetCore.Builder.RouterMiddleware.<Invoke>d__4.MoveNext()
Satellite.Host> --- End of stack trace from previous location where exception was thrown ---
Satellite.Host>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
Satellite.Host>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Satellite.Host>    at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.<Invoke>d__7.MoveNext()
Satellite.Host> dbug: Microsoft.AspNetCore.Server.Kestrel[9]
```
In [this line](https://github.com/robward-ms/WebApi/blob/NetCore-Latest/src/Microsoft.AspNet.OData.Shared/Routing/Conventions/AttributeRoutingConvention.cs#L70) requested route data use [before init](https://github.com/aspnet/Routing/blob/032bcf43b2cefe641fc6ee9ef3ab0769024a182c/src/Microsoft.AspNetCore.Routing/RouterMiddleware.cs#L34)

**Comment by xuzhg** [comment]
@robward-ms

**Comment by robward-ms** [comment]
Yep, that code is wrong.  In AspNetCore, the RouteData is present on the context only AFTER the routing is complete; until; then it's only available on the routing context. 

The fix is for Microsoft.AspNet.OData.Routing.Conventions.AttributeRoutingConvention.SelectActionImpl to get RouteData from the controller context instead of the request; after that, Microsoft.AspNet.OData.Adapters.WebApiRequestMessage.get_RouteData() gets deleted because it is no longer used.

I'll include this fix in the next PR.

**Comment by robward-ms** [comment]
Fixed by #1138 .
